sound/soc/sof/intel/hda-dai.c

```diff
@@ -169,19 +169,28 @@ static int hda_link_pcm_prepare(struct snd_pcm_substream *substream,
 				struct snd_soc_dai *dai)
 {
 	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dpcm *dpcm;
 	struct snd_sof_dev *sdev =
 		snd_soc_component_get_drvdata(dai->component);
 	struct snd_sof_pcm *spcm;
 	int stream = substream->stream;
 
-	spcm = snd_sof_find_spcm_dai(sdev, rtd);
-	if (!spcm)
-		return -EINVAL;
+	for_each_dpcm_fe(rtd, stream, dpcm) {
+		spcm = snd_sof_find_spcm_dai(sdev, dpcm->fe);
+		if (!spcm)
+			return -EINVAL;
 
```

Comment:
@ranj063 I'm just wondering in which case spcm will be NULL. I believe the initial intent of prepare() is to set link DMA hw_params when system returns from S3. If spcm can be NULL, does it mean that link DMA register is not restored 100%? So my concern is whether it will impact on the S3 stress test? Or we can make sure that every time return from S3, spcm is always valid?

Reply:
Or use continue, not return? Any suggestion? @plbossart @xiulipan

Reply:
@libinyang if there is a FE connected to a BE dai, there will always be a valid spcm. But I can't use snd_sof_find_spcm_dai() without checking. 

ANd I dont think -ENODEV is the correct return value. -EINVAL is more apprppriate.

Reply:
@ranj063 If the prepare() is called, the BE we are using should alwasy connect to a FE. So we don't need to have a check?

> @libinyang if there is a FE connected to a BE dai, there will always be a valid spcm. But I can't use snd_sof_find_spcm_dai() without checking.

Reply:
@libinyang sorry but if snd_sof_find_spcm_dai() can return NULL, the return value has to be checked.

Reply:
@ranj063 OK. If my understanding is right, is it saying that the NULL check is just for sanity check? Or we should use "continue" instread of "return -EINVAL"?

Reply:
@ranj063 What I concern is if "return -EINVAL“ is really executed in some case, does it mean the link DMA register will not be restored when resume from S3.

Reply:
@libinyang yes, if it is invalid, prepare will fail and the link dam will not resume properly. But that should be expected isnt it? I dont understand your point.

Reply:
Link DMA not being restored properly means resume from S3 when playback will have issue just like we met before.

Reply:
@libinyang you are making this review unnecessarily complicated. If there is an error, we cannot continue. A program does not guarantee it will always succeed. But it should be capture all possible errors and take actions appropriately. If there is not spcm for a FE dai, it is an error and we should quit at this point.

Reply:
@ranj063 After a second thought, maybe I did think too complicated. Mostly spcm should not be NULL. If it's NULL, it really means there is something wrong. So we will not grant it works properly. And we currently we only connect to one FE for a BE. So 'continue' and 'return error' directly should be the same. I'm OK with this patch now. If spcm is NULL, just let it be wrong and report it.